testcontainers/testcontainers-dotnet
# feat: Add Apache Pulsar module

## What does this PR do?

This PR introduces Apache Pulsar support to testcontainers-dotnet, enabling users to easily spin up Pulsar containers for testing purposes. It also provides an option to enable TokenAuthentication, allowing for secure Pulsar interactions.

## Why is it important?

Apache Pulsar is a rapidly growing distributed messaging system gaining traction in the cloud-native ecosystem. Integrating Pulsar support into testcontainers-dotnet will make it easier for developers to test their applications using Pulsar. Additionally, providing TokenAuthentication support enhances security and privacy in Pulsar-based testing scenarios.

## How to test this PR

Run the XUnit tests in the testcontainers-dotnet project to ensure the Pulsar containerization and TokenAuthentication functionality work as expected.

## Follow-ups

@HofmeisterAn, I encountered a challenge while adding [WithTokenAuthentication ](https://github.com/entvex/testcontainers-dotnet/blob/db9156d6707a00bace05994b637a0a7647085e58/src/Testcontainers.Pulsar/PulsarBuilder.cs#L71)capabilities to PulsarBuilder. The current implementation of WithCommand appends commands instead of replacing them, making it difficult to modify container startup commands based on selected Pulsar features. In the testcontainers-java module, the container command is modified just before the container starts, enabling control over its execution. Could you provide suggestions on how to achieve a similar approach in testcontainers-dotnet?

In the java module they seem to have the control to edit it, [just before the container starts](https://github.com/testcontainers/testcontainers-java/blob/78d17685359de57dd770d3dbf5e24433d40698fb/modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java#L96)

 Do you have a suggestion on a suggestion on how I should implement this for testcontainers-dotnet ?
